core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,237 @@
+}
+
+class AvroUpsertMonitorActor(
+                                  val pollInterval: FiniteDuration,
+                                  pollIntervalJitter: FiniteDuration,
+                                  workspaceService: UserInfo => WorkspaceService,
+                                  googleServicesDAO: GoogleServicesDAO,
+                                  samDAO: SamDAO,
+                                  googleStorage: GoogleStorageService[IO],
+                                  pubSubDao: GooglePubSubDAO,
+                                  pubSubSubscriptionName: String,
+                                  avroUpsertBucketName: String,
+                                  batchSize: Int)(implicit cs: ContextShift[IO])
+  extends Actor
+    with LazyLogging
+    with FutureSupport {
+  import AvroUpsertMonitor._
+  import context._
+
+  case class AvroMetadataJson(namespace: String, name: String, userSubjectId: String, userEmail: String)
+  implicit val avroMetadataJsonFormat = jsonFormat4(AvroMetadataJson)
+
+  self ! StartMonitorPass
+
+  // fail safe in case this actor is idle too long but not too fast (1 second lower limit)
+  setReceiveTimeout(max((pollInterval + pollIntervalJitter) * 10, 1 second))
+
+  private def max(durations: FiniteDuration*): FiniteDuration = {
+    implicit val finiteDurationIsOrdered = scala.concurrent.duration.FiniteDuration.FiniteDurationIsOrdered
+    durations.max
+  }
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      logger.info(s"received avro upsert message: $message")
+      val (jobId, file) = parseMessage(message)
+
+      file match {
+        case "upsert.json" => initAvroUpsert(jobId, file).map(response => (response, message.ackId)) pipeTo self
+        case _ => acknowledgeMessage(message.ackId) pipeTo self //some other file (i.e. success/failure log, metadata.json)
+      }
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime.asInstanceOf[FiniteDuration], self, StartMonitorPass)
+
+    case ((), ackId: String) => self ! None
+      acknowledgeMessage(ackId).map(_ => StartMonitorPass) pipeTo self
+//
+    case Status.Failure(t) => throw t
+
+    case ReceiveTimeout =>
+      throw new WorkbenchException("AvroUpsertMonitorActor has received no messages for too long")
+
+    case x =>
+      logger.info(s"unhandled $x")
+      self ! None
+  }
+
+  private def initAvroUpsert(jobId: String, file: String): Future[Unit] = {
+    for {
+      avroUpsertJson <- readUpsertObject(s"$jobId/upsert.json")
```

Comment:
If this future is created outside of the for comprehension and this line moved to right above the upsert loop, it can be processed in parallel with reading the metadata and getting the pet. Might be a win because this read could take some time and getting the pet involves some network calls.

Reply:
Good idea- done!